torchmultimodal/models/gpt.py

```diff
@@ -0,0 +1,405 @@
+        causal (bool. optional): If ``True``, use causal attention. Defaults to ``False``.
+        return_attn_weights (bool, optional): If ``True``, returns attention probabilities of the layer.
+            Defaults to ``False``.
+    """
+
+    def __init__(
+        self,
+        d_model: int = 768,
+        n_head: int = 12,
+        dropout: float = 0.1,
+        activation: Callable[..., nn.Module] = SiLU,
+        attn_module: nn.Module = SelfAttention(attn_dropout=0.1),
+    ) -> None:
+        super().__init__()
+
+        self.norm_attn = nn.LayerNorm(d_model)
+        self.norm_mlp = nn.LayerNorm(d_model)
+        self.dropout_attn = nn.Dropout(dropout)
+        self.dropout_mlp = nn.Dropout(dropout)
+
+        self.attention = MultiHeadAttention(
+            dim_q=d_model,
+            dim_kv=d_model,
+            n_head=n_head,
+            attn_module=attn_module,
+        )
+
+        self.mlp = MLP(
+            in_dim=d_model,
+            out_dim=d_model,
+            hidden_dims=[d_model * 4],
+            dropout=0.0,
+            activation=activation,
+        )
+
+    def forward(
+        self,
+        x: Tensor,
+        attn_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+        use_cache: bool = False,
+        causal: bool = False,
+        return_attn_weights: bool = False,
+    ) -> TransformerLayerOutput:
+        attn_probs = None
+        past_key_values = None
+
+        # Add head dim to attention mask for broadcasting
+        if attn_mask is not None:
+            attn_mask = get_extended_attention_mask(attn_mask)
+
+        attn_out = self._attn(
+            self.norm_attn(x),
+            attn_mask,
+            head_mask,
+            return_attn_weights,
+            use_cache=use_cache,
+            causal=causal,
+        )
+
+        if return_attn_weights:
+            x, attn_probs = attn_out
+        else:
+            x = attn_out
+
+        if use_cache:
+            past_key_values = self.attention.cache
+
+        x = x + self.dropout_attn(x)
+        x = self._mlp_block(self.norm_mlp(x))
+        x = x + self.dropout_mlp(x)
+
+        return TransformerLayerOutput(
+            hidden_states=x,
+            attention_weights=attn_probs,
+            past_key_values=past_key_values,
+        )
+
+    @checkpoint_wrapper
+    def _attn(
+        self,
+        x: Tensor,
+        attn_mask: Tensor,
+        head_mask: Tensor,
+        return_attn_weights: bool,
+        use_cache: bool,
+        causal: bool,
+    ) -> Union[Tensor, Tuple[Tensor, Tensor]]:
+        return self.attention(
+            x,
+            attention_mask=attn_mask,
+            head_mask=head_mask,
+            return_attn_weights=return_attn_weights,
+            use_cache=use_cache,
+            causal=causal,
+        )
+
+    @checkpoint_wrapper
+    def _mlp_block(self, x: Tensor) -> Tensor:
+        return self.mlp(x)
+
+
+class RightShift(nn.Module):
```

Comment:
Where else is this used?

Reply:
Not in this PR but later:
- This needs to be part of training as `sos` is a model parameter. 
- The trained `sos` will be the first input token during generation.

Context:
- MUGEN [inserts](https://github.com/mugen-org/MUGEN_baseline/blob/main/lib/models/gpt/attention.py#L84) `RightShift` in the middle of GPT after position embedding and before transformer decoder. 
- RightShift only gets triggered in the two modes I mentioned above and their implementation IMO is quite [hacky](https://github.com/mugen-org/MUGEN_baseline/blob/main/lib/models/gpt/attention.py#L318).

My plan:
- Trigger `RightShift` when `self.training` is `True` such that I don't have to break apart GPT into position embedding + RightShift + TransformerDecoder.
- During generation load the trained `sos` from the weight file to initialize the sequence to be generated